### PR TITLE
[Core] [Frontend] Make detokenization optional

### PR DESCRIPTION
**Pull request by mgerstgrasser**
FIX #3635 

This PR makes detokenization optional during generation. The changes here are the minimal ones required to enable this: We add an optional `detokenize` argument to `SamplingParams` which defaults to `True`. In `LLMEngine` we then skip detokenization if `detokenize` is set to `False` for the current request. In that case, `CompletionOutput.text` is returned empty, but `CompletionOutput.token_ids` contains the generated token IDs.

One thing I don't know is how (if at all) we could make this work with th OpenAI API server, at least in combination with the `openai` python client library. If anyone has ideas on that, or of course any other feedback, I'd be happy to make changes!

### CODE REVIEW

**Comment by mgerstgrasser** [comment]
Quick code snippet to test this:
```
from vllm import LLM, SamplingParams

llm = LLM(
    model=f"gpt2",
)

prompt = "Today is a good day to"

print("Now with detokenize=False")
sampling_params = SamplingParams(max_tokens=10, temperature=0.0, detokenize=False)
outputs = llm.generate(prompt, sampling_params)
print(outputs[0].outputs[0])

print("Now with detokenize=True")
sampling_params = SamplingParams(max_tokens=10, temperature=0.0, detokenize=True)
outputs = llm.generate(prompt, sampling_params)
print(outputs[0].outputs[0])
```
yields (skipping the tqdm bar):
```
Now with detokenize=False
CompletionOutput(index=0, text='', token_ids=[307, 257, 636, 286, 262, 995, 13, 198, 198, 40], cumulative_logprob=-16.733878153841943, logprobs=None, finish_reason=length, stop_reason=None)
Now with detokenize=True
CompletionOutput(index=0, text=' be a part of the world.\n\nI', token_ids=[307, 257, 636, 286, 262, 995, 13, 198, 198, 40], cumulative_logprob=-16.733878153841943, logprobs=None, finish_reason=length, stop_reason=None)
```

With a larger list of prompts and larger `max_tokens` I see a roughly 20% speedup on an A100 when disabling detokenization.

**Comment by ywang96** [comment]
+1 on this feature - in practice this feature can benefit a lot of use cases when the detokenization is not necessary and only the token ids themselves are needed for downstream tasks.

We also observe roughly 10-15% speedup on `ShareGPT` when simply commenting out
https://github.com/vllm-project/vllm/blob/9c82a1bec3a177ff2d611c092c19d25cabd90bb0/vllm/engine/llm_engine.py#L419-L421 

There's also https://github.com/vllm-project/vllm/pull/3748, so I think we should consolidate the efforts for this feature.

**Comment by mgerstgrasser** [comment]
After digging some more, I don't see an easy way to support this through the OpenAI server, at least while maintaining close compatibility wih their API. Instead, I propose adding token input and output to the (non-OpenAI) API server. I've added a proposed implementation to this PR. The API server now takes either `prompt` or `prompt_token_ids` as input. If `detokenize` is set to `True` it will return text, otherwise it will return `token_ids`.

**Comment by mgerstgrasser** [comment]
> +1 on this feature - in practice this feature can benefit a lot of use cases when the detokenization is not necessary and only the token ids themselves are needed for downstream tasks.
> 
> We also observe roughly 10-15% speedup on `ShareGPT` by simply commenting out
> 
> https://github.com/vllm-project/vllm/blob/9c82a1bec3a177ff2d611c092c19d25cabd90bb0/vllm/engine/llm_engine.py#L419-L421
> 
> ~There's also #3748, so I think we should consolidate the efforts for this feature~ Edit: the other PR simply disables the use of tokenizer completely, so these two features should technically be able to co-exist.

Thank you! And yeah, the other PR was opened almost the same minute as mine, otherwise we could have coordinated. They are indeed somewhat orthogonal though. I don't know if removing the tokenizer altogether gives an additional performance benefit, but there may be other considerations.

**Comment by ywang96** [comment]
> After digging some more, I don't see an easy way to support this through the OpenAI server, at least while maintaining close compatibility wih their API. Instead, I propose adding token input and output to the (non-OpenAI) API server. 

@mgerstgrasser The non-OpenAI API server is actually deprecated already and no longer accepting changes. IMO if OpenAI API does not support token ids only as output, then we should keep it as a feature on the engine level instead of server level (at least for this PR).

I believe the goal of the OpenAI API server is the ability to serve an OSS model as a drop-in placement for OpenAI endpoints, so I think we should try to include as little vLLM custom logic there as possible. Plus, there's no restriction for users to build their own API servers with just `AsyncLLMEngine`. WDYT?

**Comment by mgerstgrasser** [comment]
> > After digging some more, I don't see an easy way to support this through the OpenAI server, at least while maintaining close compatibility wih their API. Instead, I propose adding token input and output to the (non-OpenAI) API server.
> 
> @mgerstgrasser The non-OpenAI API server is actually deprecated already and no longer accepting changes. IMO if OpenAI API does not support token ids only as output, then we should keep it as a feature on the engine level instead of server level (at least for this PR).
> 
> I believe the goal of the OpenAI API server is the ability to serve an OSS model as a drop-in placement for OpenAI endpoints, so I think we should try to include as little vLLM custom logic there as possible. Plus, there's no restriction for users to build their own API servers with just `AsyncLLMEngine`. WDYT?

Yes, absolutely! I kind of did the API server changes for my own project primarily (so exactly as you suggest, easy to build your own API server if needed). I figured I'd be happy to share that, but also don't mind removing it from the PR!

**Comment by ywang96** [comment]
@mgerstgrasser Hello! Thanks for making this PR - IMO overall the logic looks good to me, but do you mind adding some tests just as a sanity check?

The other thing I'm not sure about is whether we should enforce `detokenize=True` if the request is coming from a call to the OAI server in `serving_chat.py` and `serving_completion.py`, since the OpenAI API itself does not support token ids only as output.

**Comment by ywang96** [comment]
> Yes, absolutely! Is there anywhere existing where these could fit in well? I don't see any end-to-end tests where I could easily add something.

Perhaps a new file under `tests/engine`?


> Since CompletionRequest and ChatCompletionRequest don't include any logic to override it

Oh Yes that's right - could you leave a `# NOTE` like the following to indicate that this parameter is only exposed at the engine level for now? Thanks!https://github.com/vllm-project/vllm/blob/b95047f2dad3cc4d64e6f4d76974e3c55917c5ee/vllm/utils.py#L141-L143

**Comment by mgerstgrasser** [comment]
> > Yes, absolutely! Is there anywhere existing where these could fit in well? I don't see any end-to-end tests where I could easily add something.
> 
> Perhaps a new file under `tests/engine`?
> 
> > Since CompletionRequest and ChatCompletionRequest don't include any logic to override it
> 
> Oh yes that's right - could you leave a `# NOTE` like the following to indicate that this parameter is only exposed at the engine level for now? Thanks!
> 
> https://github.com/vllm-project/vllm/blob/b95047f2dad3cc4d64e6f4d76974e3c55917c5ee/vllm/utils.py#L141-L143

Done, and done! I see a few checks failed now, but as far as I can tell due to internal errors?

File: vllm/sampling_params.py [hunk 1]
```diff
@@ -210,6 +216,10 @@ def _verify_args(self) -> None:
         if self.prompt_logprobs is not None and self.prompt_logprobs < 0:
             raise ValueError(f"prompt_logprobs must be non-negative, got "
                              f"{self.prompt_logprobs}.")
+        if len(self.stop) > 0 and self.detokenize is False:
```
Comment (njhill) [inline vllm/sampling_params.py:219]: nit: minor simplification
```suggestion
        if self.stop and not self.detokenize:
```

**Review by njhill — APPROVED** [review]
Looks great to me thanks @mgerstgrasser @ywang96

**Comment by njhill** [comment]
@mgerstgrasser looks like you need to run `format.sh` too

**Comment by mgerstgrasser** [comment]
> @mgerstgrasser looks like you need to run `format.sh` too

@njhill I did, not sure why it failed earlier, but the simplification commit triggered a re-run and seems to be good now.

**Review by njhill — APPROVED** [review]

**Review by ywang96 — APPROVED** [review]
@mgerstgrasser Thank you again for working on this PR!

**Comment by mgerstgrasser** [comment]
> @mgerstgrasser Thank you again for working on this PR!

Of course! Thank you both @ywang96 @njhill for being so quick and responsive with reviewing, and for all your work on this fantastic project in general!

**Pull request merged**